utils/src/main/scala/za/co/absa/enceladus/utils/schema/SchemaUtils.scala

```diff
@@ -116,14 +115,59 @@ object SchemaUtils {
   }
 
   /**
-    * Get first array column's path out of complete path.
-    *
-    *  E.g if the path argument is "a.b.c.d.e" where b and d are arrays, "a.b" will be returned.
-    *
-    * @param path   The path to the attribute
-    * @param schema The schema of the whole dataset
-    * @return The path of the first array field or "" if none were found
-    */
+   * Returns all renames in the provided schema.
+   * @param schema  schema to examine
+   * @return        the keys of the returned map are the columns' names after renames, the values are the source columns;
+   *                the name are full paths denoted with dot notation
+   */
+  def getRenamesInSchema(schema: StructType): Map[String, String] = {
+
+    def getRenamesRecursively(path: String, struct: StructType, acc: Map[String, String]): Map[String, String] = {
+      import za.co.absa.enceladus.utils.implicits.StructFieldImplicits.StructFieldEnhancements
+
+      struct.fields.foldLeft(acc) { (renames, field) =>
+        val fieldPath = appendPath(path, field.name)
+        // fieldOwnRename will be a map of only one item only if sourcecolumn is present in metadata and differs from field.name
+        // otherwise fieldOwnRename will be empty
+        val fieldOwnRename  = field.getMetadataString(MetadataKeys.SourceColumn)
+          .filter(_ != field.name)
+          .map(inputName => (fieldPath, appendPath(fieldPath, inputName)))
+          .toMap
+        val renamesWithLocal = renames ++ fieldOwnRename
+
+        field.dataType match {
+          case st: StructType => getRenamesRecursively(fieldPath, st, renamesWithLocal)
+          case at: ArrayType  =>
+            getStructInArray(at.elementType)
+              .map(x => getRenamesRecursively(fieldPath, x, renamesWithLocal))
```

Comment:
[scala.Option](https://www.scala-lang.org/api/2.11.12/#scala.Option) has a method `fold` which might seem nice in this situation. It is equivalent to `$option map f getOrElse ifEmpty`

```
getStructInArray(at.elementType).fold(renamesWithLocal)(x => getRenamesRecursively(fieldPath, x, renamesWithLocal))

```